File: code_snippets/linreg.ml
open Ppl
let linreg_model points = 
  let linreg' =
    let* m = normal 0. 2. in
    let* c = normal 0. 2. in
    List.fold 
      points
      ~init:(return (m,c))
      ~f:(fun d (x,y) -> observe y (Primitive.(normal (m*x+c) 1.)) d)

let slope = fmap fst (linreg_model points)
let y_intercept = fmap snd (linreg_model points)

File: code_snippets/anglican/linreg.clj
(defquery linreg_model
  (let [m (sample (normal 0. 2.))
        c (sample (normal 0. 2.))
        observations [[0 0] [1 2]
                      [2 4] [3 6]
                      [4 8] [5 10]
                      [6 12] [7 14]]
        ]
    (map
     (fn [[x y]]
       (observe
        (normal (+ c (* m x)) 1.)
        y))
     observations)
    m
  )
)
